File: C++/25.访问者模式/ComputerPartDisplayVisitor.cpp
/**
 * @file ComputerPartDisplayVisitor.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-29
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "ComputerPartDisplayVisitor.h"

#include <iostream>

#include "Computer.h"
#include "Keyboard.h"
#include "Monitor.h"
#include "Mouse.h"

void ComputerPartDisplayVisitor::Visit(std::shared_ptr<Computer> computer) {
    std::cout << "Displaying Computer." << std::endl;
}

void ComputerPartDisplayVisitor::Visit(std::shared_ptr<Mouse> mouse) { std::cout << "Displaying Mouse." << std::endl; }

void ComputerPartDisplayVisitor::Visit(std::shared_ptr<Keyboard> keyboard) {
    std::cout << "Displaying Keyboard." << std::endl;
}

void ComputerPartDisplayVisitor::Visit(std::shared_ptr<Monitor> monitor) {
    std::cout << "Displaying Monitor." << std::endl;
}

File: Objective-C/25.VisitorPattern/ComputerPartDisplayVisitor.m
//
//  ComputerPartDisplayVisitor.m
//  25.VisitorPattern
//
//  Created by 黄磊 on 2020/11/21.
//

#import "ComputerPartDisplayVisitor.h"

@implementation ComputerPartDisplayVisitor

- (void)visitComputer:(nonnull Computer *)computer {
    NSLog(@"Displaying Computer.");
}

- (void)visitKeyboard:(nonnull Keyboard *)keyboard {
    NSLog(@"Displaying Keyboard.");
}

- (void)visitMonitor:(nonnull Monitor *)monitor {
    NSLog(@"Displaying Monitor.");
}

- (void)visitMouse:(nonnull Mouse *)mouse {
    NSLog(@"Displaying Mouse.");
}

@end
